File: glove/python/src/imu.py
from __future__ import annotations

import argparse

from qnbot_sdk import Sample, Sdk, SerialConnection, Side
from qnbot_sdk.glove import GloveConfig, GloveImu


def print_imu(sample: Sample[GloveImu]) -> None:
    imu = sample.value.payload
    print(
        f"imu sequence={sample.sequence} valid={imu.valid} "
        f"gyroscope_raw={imu.gyroscope_raw} "
        f"accelerometer_raw={imu.accelerometer_raw}"
    )


def main() -> None:
    arguments = argparse.ArgumentParser(description="Read wired glove IMU raw counts")
    arguments.add_argument("--port", required=True, help="Serial port for the glove")
    arguments.add_argument(
        "--side",
        choices=(Side.LEFT.value, Side.RIGHT.value),
        required=True,
        help="Physical glove side",
    )
    options = arguments.parse_args()

    sdk = Sdk(
        devices=(
            GloveConfig(
                side=Side(options.side),
                connection=SerialConnection(port=options.port),
            ),
        )
    )
    glove = sdk.glove()
    imu = glove.device().imu()
    _imu_subscription = imu.subscribe(print_imu)

    glove.start()
    print("running; press Ctrl+C to stop", flush=True)
    glove.run_forever()


if __name__ == "__main__":
    main()

File: examples/glove/cpp/src/imu.cpp
#include "example_support.hpp"

#include <cstdlib>
#include <iostream>

int main(int argc, char** argv) {
    try {
        const auto options = example::parse_serial_options(
            argc, argv, example::SerialExample::imu);
        qnbot::Sdk sdk(example::serial_config(options.port, *options.side));
        auto glove = sdk.glove();

        auto imu = glove.device("primary").imu();
        const auto subscription =
            imu.subscribe([](const qnbot::Sample<qnbot::GloveImu>& sample) {
                const auto& value = sample.value.payload;
                std::cout << "imu sequence=" << sample.sequence
                          << " valid=" << value.valid << " gyroscope_raw=["
                          << value.gyroscope_raw[0] << ", "
                          << value.gyroscope_raw[1] << ", "
                          << value.gyroscope_raw[2] << "] accelerometer_raw=["
                          << value.accelerometer_raw[0] << ", "
                          << value.accelerometer_raw[1] << ", "
                          << value.accelerometer_raw[2] << "]\n";
            });

        glove.start();
        std::cout << "running; press Ctrl+C to stop" << std::endl;
        glove.run_forever();
        static_cast<void>(subscription);
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}
